Allowing to clear the read-only flag (e.g. for changing the U-Boot env):

# echo 0x400 > /sys/class/mtd/mtd4/flags
# fw_setenv serverip 172.16.11.137
# echo 0x800 > /sys/class/mtd/mtd4/flags 


diff --git a/drivers/mtd/mtdcore.c b/drivers/mtd/mtdcore.c
index 453242d6cf56..2317facd24e4 100644
--- a/drivers/mtd/mtdcore.c
+++ b/drivers/mtd/mtdcore.c
@@ -142,7 +142,22 @@ static ssize_t mtd_flags_show(struct device *dev,
 
 	return snprintf(buf, PAGE_SIZE, "0x%lx\n", (unsigned long)mtd->flags);
 }
-static DEVICE_ATTR(flags, S_IRUGO, mtd_flags_show, NULL);
+static ssize_t mtd_flags_store(struct device *dev,
+			       struct device_attribute *attr,
+			       const char *buf, size_t count)
+{
+	struct mtd_info *mtd = dev_get_drvdata(dev);
+	unsigned int flags;
+	int retval;
+
+	retval = kstrtouint(buf, 0, &flags);
+	if (retval)
+		return retval;
+
+	mtd->flags = flags;
+	return count;
+}
+static DEVICE_ATTR(flags, S_IRUGO | S_IWUSR, mtd_flags_show, mtd_flags_store);
 
 static ssize_t mtd_size_show(struct device *dev,
 		struct device_attribute *attr, char *buf)
